include/aws/common/byte_buf.h

```diff
@@ -20,44 +20,55 @@
 #include <stdint.h>
 
 /**
- * Represents a length-delimited binary string or buffer.
+ * Represents a length-delimited binary string or buffer. If byte buffer points to constant
+ * memory or memory that should otherwise not be freed by this struct, set allocator to NULL
+ * and free function will be a no-op.
  */
 struct aws_byte_buf {
+    struct aws_allocator * allocator;
     uint8_t *buffer;
     size_t len;
 };
 
-static inline int aws_byte_buf_alloc(struct aws_allocator * allocator, struct aws_byte_buf * buf, size_t len) {
+static inline int aws_byte_buf_init(struct aws_allocator * allocator, struct aws_byte_buf * buf, size_t len) {
     buf->buffer = (uint8_t*)allocator->mem_acquire(allocator, len);
     if (!buf->buffer) return aws_raise_error(AWS_ERROR_OOM);
     buf->len = len;
+    buf->allocator = allocator;
     return AWS_OP_SUCCESS;
 }
 
-static inline void aws_byte_buf_free(struct aws_allocator * allocator, struct aws_byte_buf * buf) {
-    if (buf->buffer) allocator->mem_release(allocator, buf->buffer);
+static inline void aws_byte_buf_clean_up(struct aws_byte_buf * buf) {
+    if (buf->allocator && buf->buffer) buf->allocator->mem_release(buf->allocator, buf->buffer);
+    buf->allocator = NULL;
     buf->buffer = NULL;
     buf->len = 0;
 }
 
+/**
+ * For creating a byte buffer from a constant string. (With no nulls, else strlen will not work.)
+ */
 static inline struct aws_byte_buf aws_byte_buf_from_literal(const char *literal) {
     struct aws_byte_buf buf;
     buf.buffer = (uint8_t *)literal;
     buf.len = strlen(literal);
+    buf.allocator = NULL;
     return buf;
 }
 
-static inline struct aws_byte_buf aws_byte_buf_from_c_str(const char *c_str, size_t len) {
+static inline struct aws_byte_buf aws_byte_buf_from_c_str(struct aws_allocator * allocator, const char *c_str, size_t len) {
```

Comment:
why would I want an allocator here?

Reply:
I thought there might be a use case where we had already allocated the array or string and then wanted to hand over that ownership to the byte buffer and I figured it was easiest to give the caller the option of whether to pass in an allocator for deallocation or to pass in NULL. But I'll change this and the other one to just not take the allocator argument and set the allocator to NULL.